Add workflow support

BenjaminGill-Metaswitch opened the pull request:
This seems to compile, but I haven't managed to actually test this properly...

Should resolve [JENKINS-42629](https://issues.jenkins-ci.org/browse/JENKINS-42629)

~- [ ] Work out whether testable~
~- [ ] Write up necessary configuration in pipelines (declarative and legacy)~

Commit:
Add workflow support

Should resolve JENKINS-42629

Commit:
Run covers everything

Daniel15 commented:
Thanks for the pull request. I'd like to get confirmation that it actually works before merging the pull request 😃 

Some unit tests would be nice but I know very little about Java and am not sure how to write unit tests for Jenkins plugins :(

BenjaminGill-Metaswitch commented:
OK. I've now installed the new package on a production system, and it appears to work (i.e. it can now raise issues within a pipeline where before it couldn't).

I don't really understand the UT framework here, but will have a go at seeing if I can test this...

Commit:
Treat null Result as success

In pipelines, Result is only set at the end. The common pattern is to
explicitly set it to failed/unstable. Thus, we can assume that null =>
success.

BenjaminGill-Metaswitch commented:
Right. I've been using this for the past 10 days in production without significant problems. I've still not worked out how the UTs work for this plugin, though. 

I think the correct place for documentation will be https://wiki.jenkins-ci.org/display/JENKINS/GitHub+Issues+Plugin, once a release appears with this feature:

As an example (note that the GitHub project must be configured in the main Jenkins config):

```groovy
script {
    properties([[$class: 'GithubProjectProperty',
                projectUrlStr: '<GitHub repo URL>']])
}

...

step([$class: 'GitHubIssueNotifier',
      issueAppend: true,
      issueLabel: '',
      issueTitle: '$JOB_NAME $BUILD_DISPLAY_NAME failed'])
```

Is this worth merging now?

BenjaminGill-Metaswitch commented:
Is this able to be merged?

Daniel15 commented:
Sorry, I'll take a look soon. Thanks for your contribution!

Sent from my phone.

On Jul 7, 2017 5:20 AM, "Benjamin Gill" <notifications@github.com> wrote:

> Is this able to be merged?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/github-issues-plugin/pull/8#issuecomment-313666296>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFnHQVI57AGNx6i5_jvOPiG2uBrmdehks5sLiJ-gaJpZM4Nugxs>
> .
>

Daniel15 commented on src/main/java/org/jenkinsci/plugins/githubissues/GitHubIssueNotifier.java:
```diff
@@ -181,15 +182,15 @@ public void perform(
                 run.addAction(new GitHubIssueAction(issue, GitHubIssueAction.TransitionAction.OPEN));
             }
-        } else if (result == Result.SUCCESS && issue != null && issue.getState() == GHIssueState.OPEN) {
+        } else if ((result == Result.SUCCESS || result == null) && issue != null && issue.getState() == GHIssueState.OPEN) {
```
Why do you need to add `result == null`?

BenjaminGill-Metaswitch replied:
In declarative pipelines, the result will be `null` unless explicitly set. The common pattern is to set the failure state within a `catch` block, so we can treat `null` as `SUCCESS`. I've added an explanatory comment.

Daniel15 replied:
Thank you for the explanation :)

Commit:
Explain `result == null`

Daniel15 commented:
Published as v1.2.3. Thanks for your contribution!
